plugins/gpio/fu-gpio-device.c

```diff
@@ -0,0 +1,206 @@
+		fu_device_add_instance_id(device, instance_id);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_gpio_device_unassign(FuGpioDevice *self, GError **error)
+{
+	if (self->fd < 0)
+		return TRUE;
+	if (!g_close(self->fd, error))
+		return FALSE;
+	self->fd = -1;
+	return TRUE;
+}
+
+static gboolean
+fu_gpio_device_assign_full(FuGpioDevice *self, guint64 line, gboolean value, GError **error)
+{
+	const gchar consumer[] = "fwupd";
+	struct gpio_v2_line_request req = {
+	    .num_lines = 1,
+	    req.offsets[0] = line,
+	    .config.flags = GPIO_V2_LINE_FLAG_OUTPUT,
+	    .config.num_attrs = 1,
+	    .config.attrs[0].attr.values = value ? 0x1 : 0x0,
+	    .config.attrs[0].mask = 0x1,
+	};
+
+	/* this is useful if we have contention with other tools */
+	if (!fu_memcpy_safe((guint8 *)req.consumer,
+			    sizeof(req.consumer),
+			    0x0, /* dst */
+			    (const guint8 *)consumer,
+			    sizeof(consumer),
+			    0x0, /* src */
+			    sizeof(consumer),
+			    error))
+		return FALSE;
+
+	/* slightly weird API, but roll with it */
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  GPIO_V2_GET_LINE_IOCTL,
+				  (guint8 *)&req,
+				  NULL,
+				  error)) {
+		g_prefix_error(error, "failed to assign: ");
+		return FALSE;
+	}
+
+	/* success */
+	self->fd = req.fd;
+	return TRUE;
+}
+
+gboolean
+fu_gpio_device_assign(FuGpioDevice *self, const gchar *id, gboolean value, GError **error)
+{
+	guint64 line = G_MAXUINT64;
+
+	/* sanity check */
+	if (self->fd > 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_FOUND,
+			    "GPIO %s already in use",
+			    id);
+		return FALSE;
+	}
+
+	/* specified as a number, or look for @id as named pin */
+	if (fu_common_strtoull_full(id, &line, 0, self->num_lines - 1, NULL)) {
+		struct gpio_v2_line_info info = {.offset = line};
+		if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+					  GPIO_V2_GET_LINEINFO_IOCTL,
+					  (guint8 *)&info,
+					  NULL,
+					  error)) {
+			g_prefix_error(error, "failed to get lineinfo: ");
+			return FALSE;
+		}
+	} else {
+		for (guint i = 0; i < self->num_lines; i++) {
+			struct gpio_v2_line_info info = {.offset = i};
+			g_autofree gchar *name = NULL;
+			if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+						  GPIO_V2_GET_LINEINFO_IOCTL,
+						  (guint8 *)&info,
+						  NULL,
+						  error)) {
+				g_prefix_error(error, "failed to get lineinfo: ");
+				return FALSE;
+			}
+			name = fu_common_strsafe(info.name, sizeof(info.name));
+			if (g_strcmp0(name, id) == 0) {
+				line = i;
+				break;
+			}
+		}
+	}
+	if (line == G_MAXUINT64) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "failed to find %s", id);
+		return FALSE;
+	}
+	return fu_gpio_device_assign_full(self, line, value, error);
+}
+
+static void
+fu_gpio_device_init(FuGpioDevice *self)
+{
+}
+
+static void
+fu_gpio_device_finalize(GObject *object)
+{
+	FuGpioDevice *self = FU_GPIO_DEVICE(object);
+	if (self->fd > 0)
+		g_close(self->fd, NULL);
```

Comment:
Just to double check we're not digging a blocker hole - can you make sure that gpio-info still works while fwupd holds the `fd` for the duration of the daemon being opened?

Reply:
yup, it works fine, and you get the "consumer" shown as "fwupd" for the lifetime of the update.